File: verman-py/version_manager.py
"""
Core version management logic.
"""

import threading
import time
from typing import Dict, List, Optional, Tuple

from database import DatabaseManager
from file_manager import FileManager
from logger import logger as app_logger
from models import (
    ACTIVE_FILE_STATUSES,
    CreateVersionResult,
    RollbackResult,
    ScanSnapshot,
)


class VersionManager:
    """Business logic for create, compare, rollback, export, and query."""

    def __init__(self, db_manager: DatabaseManager, file_manager: FileManager, config_manager=None):
        self.db_manager = db_manager
        self.file_manager = file_manager
        self.config_manager = config_manager
        self._operation_lock = threading.RLock()
        self._last_scan_snapshot: Optional[ScanSnapshot] = None
        self._last_scan_time = 0.0
        self._cache_ttl = 1.0

    def _get_ignore_patterns(self) -> List[str]:
        if self.config_manager:
            try:
                return self.config_manager.get_ignore_patterns()
            except Exception as exc:
                app_logger.warning(f"获取忽略模式失败: {exc}")
        return []

    def refresh_workspace(self, force: bool = False) -> ScanSnapshot:
        with self._operation_lock:
            current_time = time.time()
            if (
                not force
                and self._last_scan_snapshot is not None
                and current_time - self._last_scan_time < self._cache_ttl
            ):
                return self._last_scan_snapshot

            ignore_patterns = self._get_ignore_patterns()
            indexed_files = self.db_manager.get_workspace_index()
            current_files, blocked_files = self.file_manager.scan_workspace(ignore_patterns, indexed_files)
            scan_id = time.time_ns()
            self.db_manager.update_workspace_index(current_files.values(), scan_id)

            previous_files = self._get_latest_version_hashes()
            changes = self._detect_changes(
                current_files=current_files,
                previous_files=previous_files,
                blocked_paths={blocked_file.relative_path for blocked_file in blocked_files},
            )

            snapshot = ScanSnapshot(
                current_files=current_files,
                changes=changes,
                blocked_files=blocked_files,
                scan_id=scan_id,
            )
            self._last_scan_snapshot = snapshot
            self._last_scan_time = current_time
            return snapshot

    def get_current_changes(self) -> List[Dict]:
        try:
            return self.refresh_workspace().changes
        except Exception as exc:
            app_logger.error(f"获取变更失败: {exc}")
            return []

    def create_version(
        self,
        description: str,
        scan_snapshot: Optional[ScanSnapshot] = None,
    ) -> CreateVersionResult:
        with self._operation_lock:
            try:
                snapshot = scan_snapshot or self.refresh_workspace(force=True)
                if not snapshot.changes:
                    return CreateVersionResult(success=False, error="没有检测到文件变更。")

                app_logger.info("开始创建版本...")
                previous_files = self._get_latest_version_hashes()
                version_files = self._prepare_version_files(
                    current_files=snapshot.current_files,
                    previous_files=previous_files,
                    changes=snapshot.changes,
                )

                version_number = self._generate_version_number()
                version_id = self.db_manager.create_version(
                    version_number=version_number,
                    description=description,
                    change_count=len(snapshot.changes),
                )
                self._save_files_in_batches(version_id, version_files)
                self._clear_scan_cache()

                app_logger.info(
                    f"版本 {version_number} 创建成功，包含 {len(snapshot.changes)} 个变更"
                )
                return CreateVersionResult(
                    success=True,
                    version_number=version_number,
                    change_count=len(snapshot.changes),
                )
            except Exception as exc:
                app_logger.error(f"创建版本失败: {exc}")
                return CreateVersionResult(success=False, error=str(exc))

    def rollback_to_version(self, version_id: int, backup_current: bool = True) -> RollbackResult:
        with self._operation_lock:
            try:
                version_files = self._get_effective_version_files(version_id)
                if not version_files:
                    return RollbackResult(success=False, error="版本文件不存在。")

                ignore_patterns = self._get_ignore_patterns()
                restore_result = self.file_manager.restore_files(
                    version_files,
                    ignore_patterns=ignore_patterns,
                    backup_current=backup_current,
                )
                is_valid, warnings = self._verify_rollback_result(version_files)
                combined_warnings = restore_result["warnings"] + warnings
                if not is_valid:
                    return RollbackResult(
                        success=False,
                        restored_count=int(restore_result["restored_count"]),
                        removed_count=int(restore_result["removed_count"]),
                        warnings=combined_warnings,
                        error="回滚结果验证失败。",
                    )

                self._clear_scan_cache()
                app_logger.info(f"成功回滚到版本 {version_id}")
                return RollbackResult(
                    success=True,
                    restored_count=int(restore_result["restored_count"]),
                    removed_count=int(restore_result["removed_count"]),
                    warnings=combined_warnings,
                )
            except Exception as exc:
                app_logger.error(f"回滚失败: {exc}")
                return RollbackResult(success=False, error=str(exc))

    def get_all_versions(self) -> List[Dict]:
        try:
            return self.db_manager.get_all_versions()
        except Exception as exc:
            app_logger.error(f"获取版本列表失败: {exc}")
            return []

    def get_version_details(self, version_id: int) -> Optional[Dict]:
        try:
            versions = self.db_manager.get_all_versions()
            version_info = next((version for version in versions if version["id"] == version_id), None)
            if not version_info:
                return None

            files = self._get_effective_version_files(version_id)
            add_count = len([file_info for file_info in files if file_info["file_status"] == "add"])
            modify_count = len(
                [file_info for file_info in files if file_info["file_status"] == "modify"]
            )
            delete_count = len([file_info for file_info in files if file_info["file_status"] == "delete"])
            unmodified_count = len(
                [file_info for file_info in files if file_info["file_status"] == "unmodified"]
            )

            version_info["files"] = files
            version_info["statistics"] = {
                "add_count": add_count,
                "modify_count": modify_count,
                "delete_count": delete_count,
                "unmodified_count": unmodified_count,
                "total_count": len(files),
            }
            return version_info
        except Exception as exc:
            app_logger.error(f"获取版本详情失败: {exc}")
            return None

    def compare_versions(self, version_id1: int, version_id2: int) -> Dict:
        try:
            files1 = self._get_effective_version_files(version_id1, include_content=False)
            files2 = self._get_effective_version_files(version_id2, include_content=False)
            return self._compare_versions_effective(files1, files2)
        except Exception as exc:
            app_logger.error(f"版本比较失败: {exc}")
            return {}

    def export_version(self, version_id: int, export_path: str) -> bool:
        try:
            version_files = self._get_effective_version_files(version_id)
            if not version_files:
                return False

            return self.file_manager.export_version_files(version_files, export_path)
        except Exception as exc:
            app_logger.error(f"导出版本失败: {exc}")
            return False

    def delete_version(self, version_id: int) -> bool:
        try:
            self.db_manager.delete_version(version_id)
            return True
        except Exception as exc:
            app_logger.error(f"删除版本失败: {exc}")
            return False

    def _get_latest_version_hashes(self) -> Dict[str, str]:
        latest_version_id = self.db_manager.get_latest_version_id()
        if latest_version_id is None:
            return {}
        return self.db_manager.get_version_file_hashes(latest_version_id)

    def _detect_changes(
        self,
        current_files: Dict[str, object],
        previous_files: Dict[str, str],
        blocked_paths: Optional[set] = None,
    ) -> List[Dict]:
        blocked_paths = blocked_paths or set()
        changes = []
        current_set = set(current_files.keys())
        previous_set = set(previous_files.keys())

        for file_path in sorted(current_set - previous_set):
            if file_path.startswith(".verman"):
                continue
            changes.append(
                {
                    "relative_path": file_path,
                    "file_hash": current_files[file_path].file_hash,
                    "file_status": "add",
                }
            )

        for file_path in sorted(current_set & previous_set):
            current_hash = current_files[file_path].file_hash
            if current_hash != previous_files[file_path]:
                changes.append(
                    {
                        "relative_path": file_path,
                        "file_hash": current_hash,
                        "file_status": "modify",
                    }
                )

        for file_path in sorted(previous_set - current_set):
            if file_path.startswith(".verman") or file_path in blocked_paths:
                continue
            changes.append(
                {
                    "relative_path": file_path,
                    "file_hash": previous_files.get(file_path, ""),
                    "file_status": "delete",
                }
            )

        return changes

    def _prepare_version_files(
        self,
        current_files: Dict[str, object],
        previous_files: Dict[str, str],
        changes: List[Dict],
    ) -> List[Dict]:
        file_status_map = {}

        for change in changes:
            file_status_map[change["relative_path"]] = {
                "status": change["file_status"],
                "hash": change["file_hash"],
            }

        for file_path, file_state in current_files.items():
            if file_path not in file_status_map:
                if previous_files.get(file_path) == file_state.file_hash:
                    file_status_map[file_path] = {"status": "unmodified", "hash": file_state.file_hash}
                else:
                    file_status_map[file_path] = {"status": "add", "hash": file_state.file_hash}

        version_files = []
        for file_path in sorted(file_status_map.keys()):
            file_info = file_status_map[file_path]
            status = file_info["status"]
            file_data = {
                "relative_path": file_path,
                "file_hash": file_info["hash"],
                "file_status": status,
                "file_content": None,
            }

            if status in {"add", "modify"}:
                file_data["file_content"] = self.file_manager.read_relative_file(file_path)

            version_files.append(file_data)

        return version_files

    def _save_files_in_batches(self, version_id: int, version_files: List[Dict]):
        batch_size = 100
        for index in range(0, len(version_files), batch_size):
            batch = version_files[index : index + batch_size]
            self.db_manager.save_files(version_id, batch, replace_existing=(index == 0))

    def _get_effective_version_files(
        self, version_id: int, include_content: bool = True
    ) -> List[Dict]:
        return self.db_manager.get_effective_version_files(version_id, include_content=include_content)

    def _clear_scan_cache(self):
        self._last_scan_snapshot = None
        self._last_scan_time = 0.0
        self.file_manager.clear_hash_cache()

    def _verify_rollback_result(self, expected_files: List[Dict]) -> Tuple[bool, List[str]]:
        try:
            ignore_patterns = self._get_ignore_patterns()
            current_files, _ = self.file_manager.scan_workspace(ignore_patterns, {})
            current_paths = set(self.file_manager.list_workspace_files(ignore_patterns))

            expected_file_map = {
                file_info["relative_path"]: file_info
                for file_info in expected_files
                if file_info["file_status"] in ACTIVE_FILE_STATUSES
            }

            mismatches = []
            for file_path, file_info in expected_file_map.items():
                current_state = current_files.get(file_path)
                if current_state is None:
                    mismatches.append(f"文件 {file_path} 未找到")
                    continue

                current_hash = current_state.file_hash
                expected_hash = file_info["file_hash"]
                if current_hash != expected_hash:
                    mismatches.append(
                        f"文件 {file_path} 哈希值不匹配: 期望 {expected_hash[:8]}, 实际 {current_hash[:8]}"
                    )

            extra_files = sorted(current_paths - set(expected_file_map.keys()))
            mismatches.extend([f"额外文件 {file_path}" for file_path in extra_files[:10]])

            if mismatches:
                app_logger.warning(f"回滚验证发现 {len(mismatches)} 个问题: {'; '.join(mismatches)}")
                return False, mismatches

            return True, []
        except Exception as exc:
            app_logger.warning(f"验证回滚结果时出错: {exc}")
            return False, [str(exc)]

    def _compare_versions_effective(self, files1: List[Dict], files2: List[Dict]) -> Dict:
        differences = {"only_in_first": [], "only_in_second": [], "different": []}

        files1_map = {
            file_info["relative_path"]: file_info
            for file_info in files1
            if file_info["file_status"] != "delete"
        }
        files2_map = {
            file_info["relative_path"]: file_info
            for file_info in files2
            if file_info["file_status"] != "delete"
        }

        paths1 = set(files1_map.keys())
        paths2 = set(files2_map.keys())

        for path in sorted(paths1 - paths2):
            differences["only_in_first"].append(files1_map[path])

        for path in sorted(paths2 - paths1):
            differences["only_in_second"].append(files2_map[path])

        for path in sorted(paths1 & paths2):
            file1 = files1_map[path]
            file2 = files2_map[path]
            if (
                file1["file_hash"] != file2["file_hash"]
                or file1["file_status"] != file2["file_status"]
            ):
                differences["different"].append(
                    {
                        "relative_path": path,
                        "file_in_v1": {
                            "file_hash": file1["file_hash"],
                            "file_status": file1["file_status"],
                        },
                        "file_in_v2": {
                            "file_hash": file2["file_hash"],
                            "file_status": file2["file_status"],
                        },
                    }
                )

        return differences

    def _generate_version_number(self) -> str:
        try:
            versions = self.db_manager.get_all_versions()
            if not versions:
                return "v1.0"

            latest_version = versions[0]["version_number"]
            if latest_version.startswith("v"):
                try:
                    major_text, minor_text = latest_version[1:].split(".")
                    new_version = f"v{int(major_text)}.{int(minor_text) + 1}"
                    if new_version not in {version["version_number"] for version in versions}:
                        return new_version
                except ValueError:
                    pass

            base_time = time.strftime("%Y%m%d_%H%M%S")
            candidate = f"v{base_time}"
            existing_versions = {version["version_number"] for version in versions}
            counter = 1
            while candidate in existing_versions:
                candidate = f"v{base_time}_{counter}"
                counter += 1
            return candidate
        except Exception:
            return f"v{int(time.time())}"

File: verman-rust/src-tauri/src/version_manager.rs
use std::collections::{HashMap, HashSet};
use std::path::Path;
use std::sync::Mutex;
use std::time::{SystemTime, UNIX_EPOCH};

use crate::database::DatabaseManager;
use crate::file_manager::FileManager;
use crate::models::{
    BLOB_STORAGE_THRESHOLD, ChangeEntry, CreateVersionResult, DiffEntry, DiffFileEntry, DiffSide,
    FileState, ProgressPayload, RollbackResult, ScanSnapshot, VersionDetails, VersionDiff,
    VersionFileInfo, VersionInfo, VersionStatistics,
};

pub struct VersionManager {
    pub db_manager: DatabaseManager,
    pub file_manager: Mutex<FileManager>,
    config_ignore_patterns: Vec<String>,
    last_scan_snapshot: Mutex<Option<ScanSnapshot>>,
    last_scan_time: Mutex<f64>,
}

#[allow(dead_code)]
impl VersionManager {
    pub fn new(
        db_manager: DatabaseManager,
        file_manager: FileManager,
        config_ignore_patterns: Vec<String>,
    ) -> Self {
        Self {
            db_manager,
            file_manager: Mutex::new(file_manager),
            config_ignore_patterns,
            last_scan_snapshot: Mutex::new(None),
            last_scan_time: Mutex::new(0.0),
        }
    }

    pub fn refresh_workspace(&self, force: bool) -> Result<ScanSnapshot, String> {
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map(|d| d.as_secs_f64())
            .unwrap_or(0.0);

        if !force {
            if let Ok(last_time) = self.last_scan_time.lock() {
                if *last_time > 0.0 && now - *last_time < 1.0 {
                    if let Ok(snapshot) = self.last_scan_snapshot.lock() {
                        if let Some(ref snap) = *snapshot {
                            return Ok(snap.clone());
                        }
                    }
                }
            }
        }

        let indexed_files = self.db_manager.get_workspace_index();

        let mut fm = self.file_manager.lock().map_err(|e| e.to_string())?;
        let (current_files, blocked_files) =
            fm.scan_workspace(&self.config_ignore_patterns, &indexed_files);
        let scan_id = now as i64 * 1_000_000_000;

        let file_states: Vec<FileState> = current_files.values().cloned().collect();
        self.db_manager
            .update_workspace_index(&file_states, scan_id)?;

        let latest_version_hashes = self.get_latest_version_hashes();
        let blocked_paths: HashSet<&str> = blocked_files
            .iter()
            .map(|b| b.relative_path.as_str())
            .collect();
        let changes = Self::detect_changes(
            &current_files,
            &latest_version_hashes,
            &blocked_paths,
        );

        let snapshot = ScanSnapshot {
            current_files,
            changes,
            blocked_files,
            scan_id,
        };

        if let Ok(mut snap) = self.last_scan_snapshot.lock() {
            *snap = Some(snapshot.clone());
        }
        if let Ok(mut t) = self.last_scan_time.lock() {
            *t = now;
        }

        Ok(snapshot)
    }

    pub fn get_current_changes(&self) -> Result<Vec<ChangeEntry>, String> {
        let snapshot = self.refresh_workspace(false)?;
        Ok(snapshot.changes)
    }

    pub fn create_version(
        &self,
        description: &str,
        scan_snapshot: Option<ScanSnapshot>,
    ) -> CreateVersionResult {
        self.create_version_with_progress(description, scan_snapshot, None)
    }

    pub fn create_version_with_progress(
        &self,
        description: &str,
        scan_snapshot: Option<ScanSnapshot>,
        on_progress: Option<&dyn Fn(ProgressPayload)>,
    ) -> CreateVersionResult {
        let snapshot = match scan_snapshot {
            Some(s) => s,
            None => match self.refresh_workspace(true) {
                Ok(s) => s,
                Err(e) => {
                    return CreateVersionResult {
                        error: Some(format!("Refresh failed: {}", e)),
                        ..Default::default()
                    }
                }
            },
        };

        if snapshot.changes.is_empty() {
            return CreateVersionResult {
                error: Some("没有检测到文件变更。".to_string()),
                ..Default::default()
            };
        }

        tracing::info!("开始创建版本...");

        let previous_hashes = self.get_latest_version_hashes();
        let version_files = self.prepare_version_files(&snapshot.current_files, &previous_hashes, &snapshot.changes);

        let version_number = self.generate_version_number();
        let version_id = match self.db_manager.create_version(
            &version_number,
            description,
            snapshot.changes.len() as i64,
        ) {
            Ok(id) => id,
            Err(e) => {
                return CreateVersionResult {
                    error: Some(e),
                    ..Default::default()
                }
            }
        };

        // Save in batches of 100
        let total_files = version_files.len();
        for (i, chunk) in version_files.chunks(100).enumerate() {
            if let Some(cb) = on_progress {
                let current = (i * 100 + chunk.len()).min(total_files);
                cb(ProgressPayload {
                    stage: "save".to_string(),
                    current,
                    total: total_files,
                    message: format!("保存中... {}/{}", current, total_files),
                });
            }
            let replace = chunk.as_ptr() == version_files.as_ptr();
            if let Err(e) = self.db_manager.save_files(version_id, chunk, replace) {
                return CreateVersionResult {
                    error: Some(e),
                    ..Default::default()
                }
            }
        }

        if let Some(cb) = on_progress {
            cb(ProgressPayload {
                stage: "done".to_string(),
                current: total_files,
                total: total_files,
                message: "版本创建完成".to_string(),
            });
        }

        self.clear_scan_cache();

        tracing::info!(version = %version_number, count = snapshot.changes.len(), "版本创建成功");

        CreateVersionResult {
            success: true,
            version_number: Some(version_number),
            change_count: snapshot.changes.len() as i64,
            ..Default::default()
        }
    }

    pub fn rollback_to_version(
        &self,
        version_id: i64,
        backup_current: bool,
    ) -> RollbackResult {
        self.rollback_to_version_with_progress(version_id, backup_current, None)
    }

    pub fn rollback_to_version_with_progress(
        &self,
        version_id: i64,
        backup_current: bool,
        on_progress: Option<&dyn Fn(ProgressPayload)>,
    ) -> RollbackResult {
        let version_files = match self
            .db_manager
            .get_effective_version_files(version_id, true)
        {
            Ok(files) => files,
            Err(e) => {
                return RollbackResult {
                    error: Some(e),
                    ..Default::default()
                }
            }
        };

        if version_files.is_empty() {
            return RollbackResult {
                error: Some("版本文件不存在。".to_string()),
                ..Default::default()
            };
        }

        if let Some(cb) = on_progress {
            cb(ProgressPayload {
                stage: "restore".to_string(),
                current: 0,
                total: version_files.len(),
                message: format!("正在恢复 {} 个文件...", version_files.len()),
            });
        }

        let fm = match self.file_manager.lock() {
            Ok(f) => f,
            Err(e) => {
                return RollbackResult {
                    error: Some(e.to_string()),
                    ..Default::default()
                }
            }
        };

        let result = fm.restore_files(
            &version_files,
            &self.config_ignore_patterns,
            backup_current,
        );

        let (restored_count, removed_count, warnings) = match result {
            Ok(r) => r,
            Err(e) => {
                return RollbackResult {
                    error: Some(e),
                    ..Default::default()
                }
            }
        };

        self.clear_scan_cache();
        tracing::info!(version_id = version_id, "成功回滚");

        if let Some(cb) = on_progress {
            cb(ProgressPayload {
                stage: "done".to_string(),
                current: version_files.len(),
                total: version_files.len(),
                message: "回滚完成".to_string(),
            });
        }

        RollbackResult {
            success: true,
            restored_count,
            removed_count,
            warnings,
            ..Default::default()
        }
    }

    pub fn get_all_versions(&self) -> Result<Vec<VersionInfo>, String> {
        self.db_manager.get_all_versions()
    }

    pub fn get_version_details(&self, version_id: i64) -> Option<VersionDetails> {
        let versions = self.db_manager.get_all_versions().ok()?;
        let version = versions.into_iter().find(|v| v.id == version_id)?;

        let files = self
            .db_manager
            .get_effective_version_files(version_id, true)
            .ok()?;

        let add_count = files.iter().filter(|f| f.file_status == "add").count();
        let modify_count = files.iter().filter(|f| f.file_status == "modify").count();
        let delete_count = files.iter().filter(|f| f.file_status == "delete").count();
        let unmodified_count = files.iter().filter(|f| f.file_status == "unmodified").count();

        Some(VersionDetails {
            id: version.id,
            version_number: version.version_number,
            create_time: version.create_time,
            description: version.description,
            change_count: version.change_count,
            files,
            statistics: VersionStatistics {
                add_count,
                modify_count,
                delete_count,
                unmodified_count,
                total_count: add_count + modify_count + delete_count + unmodified_count,
            },
        })
    }

    pub fn compare_versions(&self, id1: i64, id2: i64) -> Result<VersionDiff, String> {
        let files1 = self
            .db_manager
            .get_effective_version_files(id1, false)?;
        let files2 = self
            .db_manager
            .get_effective_version_files(id2, false)?;
        Ok(Self::compare_versions_effective(&files1, &files2))
    }

    pub fn export_version(&self, version_id: i64, export_path: &Path) -> Result<bool, String> {
        self.export_version_with_progress(version_id, export_path, None)
    }

    pub fn export_version_with_progress(
        &self,
        version_id: i64,
        export_path: &Path,
        on_progress: Option<&dyn Fn(ProgressPayload)>,
    ) -> Result<bool, String> {
        let version_files = self
            .db_manager
            .get_effective_version_files(version_id, true)?;
        if version_files.is_empty() {
            return Ok(false);
        }

        if let Some(cb) = on_progress {
            cb(ProgressPayload {
                stage: "export".to_string(),
                current: 0,
                total: version_files.len(),
                message: format!("正在导出 {} 个文件...", version_files.len()),
            });
        }

        let fm = self.file_manager.lock().map_err(|e| e.to_string())?;
        let result = fm.export_version_files(&version_files, export_path)?;

        if let Some(cb) = on_progress {
            cb(ProgressPayload {
                stage: "done".to_string(),
                current: version_files.len(),
                total: version_files.len(),
                message: "导出完成".to_string(),
            });
        }

        Ok(result)
    }

    pub fn delete_version(&self, version_id: i64) -> Result<(), String> {
        self.db_manager.delete_version(version_id)
    }

    // ── Private Helpers ──

    fn get_latest_version_hashes(&self) -> HashMap<String, String> {
        self.db_manager
            .get_latest_version_id()
            .map(|id| self.db_manager.get_version_file_hashes(id))
            .unwrap_or_default()
    }

    fn detect_changes(
        current_files: &HashMap<String, FileState>,
        previous_files: &HashMap<String, String>,
        blocked_paths: &HashSet<&str>,
    ) -> Vec<ChangeEntry> {
        let mut changes = Vec::new();
        let current_set: HashSet<&str> =
            current_files.keys().map(|s| s.as_str()).collect();
        let previous_set: HashSet<&str> =
            previous_files.keys().map(|s| s.as_str()).collect();

        // New files
        for path in difference(&current_set, &previous_set) {
            if path.starts_with(".verman") {
                continue;
            }
            if let Some(state) = current_files.get(path) {
                changes.push(ChangeEntry {
                    relative_path: path.to_string(),
                    file_hash: state.file_hash.clone(),
                    file_status: "add".to_string(),
                });
            }
        }

        // Modified files
        for path in intersection(&current_set, &previous_set) {
            if let Some(state) = current_files.get(path) {
                if let Some(prev_hash) = previous_files.get(path) {
                    if state.file_hash != *prev_hash {
                        changes.push(ChangeEntry {
                            relative_path: path.to_string(),
                            file_hash: state.file_hash.clone(),
                            file_status: "modify".to_string(),
                        });
                    }
                }
            }
        }

        // Deleted files
        for path in difference(&previous_set, &current_set) {
            if path.starts_with(".verman") || blocked_paths.contains(path) {
                continue;
            }
            changes.push(ChangeEntry {
                relative_path: path.to_string(),
                file_hash: previous_files.get(path).cloned().unwrap_or_default(),
                file_status: "delete".to_string(),
            });
        }

        changes.sort_by(|a, b| a.relative_path.cmp(&b.relative_path));
        changes
    }

    fn prepare_version_files(
        &self,
        current_files: &HashMap<String, FileState>,
        previous_hashes: &HashMap<String, String>,
        changes: &[ChangeEntry],
    ) -> Vec<VersionFileInfo> {
        let mut status_map: HashMap<&str, (&str, &str)> = HashMap::new();

        for change in changes {
            status_map.insert(
                change.relative_path.as_str(),
                (change.file_status.as_str(), change.file_hash.as_str()),
            );
        }

        for (path, state) in current_files {
            if !status_map.contains_key(path.as_str()) {
                let status = if previous_hashes.get(path) == Some(&state.file_hash) {
                    "unmodified"
                } else {
                    "add"
                };
                status_map.insert(path.as_str(), (status, state.file_hash.as_str()));
            }
        }

        let mut result: Vec<VersionFileInfo> = Vec::new();
        let mut paths: Vec<&&str> = status_map.keys().collect();
        paths.sort();

        for path in paths {
            let (status, hash) = status_map[path];
            let content = match status {
                "add" | "modify" => {
                    let fm = self.file_manager.lock().ok();
                    let full_path = fm.as_ref().map(|f| f.workspace_path().join(path));
                    let file_size = full_path.as_ref()
                        .and_then(|p| std::fs::metadata(p).ok())
                        .map(|m| m.len() as i64)
                        .unwrap_or(0);

                    if file_size > BLOB_STORAGE_THRESHOLD {
                        // Store large file as external blob
                        if let Some(ref f) = fm {
                            if let Ok(bytes) = f.read_relative_file(path) {
                                let blob_path = crate::project_paths::get_blob_path(
                                    f.workspace_path(), hash
                                );
                                std::fs::create_dir_all(blob_path.parent().unwrap()).ok();
                                std::fs::write(&blob_path, &bytes).ok();
                            }
                        }
                        None
                    } else {
                        fm.and_then(|f| f.read_relative_file(path).ok())
                    }
                }
                _ => None,
            };

            result.push(VersionFileInfo {
                relative_path: path.to_string(),
                file_hash: hash.to_string(),
                file_status: status.to_string(),
                file_content: content,
            });
        }

        result
    }

    fn generate_version_number(&self) -> String {
        let versions = match self.db_manager.get_all_versions() {
            Ok(v) => v,
            Err(_) => return format!("v{}", SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .map(|d| d.as_secs())
                .unwrap_or(0)),
        };

        if versions.is_empty() {
            return "v1.0".to_string();
        }

        let latest = &versions[0].version_number;
        if let Some(v_str) = latest.strip_prefix('v') {
            if let Some((major, minor)) = v_str.split_once('.') {
                if let (Ok(m), Ok(n)) = (major.parse::<i64>(), minor.parse::<i64>()) {
                    let candidate = format!("v{}.{}", m, n + 1);
                    if !versions.iter().any(|v| v.version_number == candidate) {
                        return candidate;
                    }
                }
            }
        }

        let base = chrono::Local::now().format("v%Y%m%d_%H%M%S").to_string();
        let existing: HashSet<&str> =
            versions.iter().map(|v| v.version_number.as_str()).collect();
        if !existing.contains(base.as_str()) {
            return base;
        }
        let mut counter = 1;
        loop {
            let candidate = format!("{}_{}", base, counter);
            if !existing.contains(candidate.as_str()) {
                return candidate;
            }
            counter += 1;
        }
    }

    fn clear_scan_cache(&self) {
        if let Ok(mut snap) = self.last_scan_snapshot.lock() {
            *snap = None;
        }
        if let Ok(mut t) = self.last_scan_time.lock() {
            *t = 0.0;
        }
        if let Ok(mut fm) = self.file_manager.lock() {
            fm.clear_hash_cache();
        }
    }

    fn compare_versions_effective(
        files1: &[VersionFileInfo],
        files2: &[VersionFileInfo],
    ) -> VersionDiff {
        let map1: HashMap<&str, &VersionFileInfo> = files1
            .iter()
            .filter(|f| f.file_status != "delete")
            .map(|f| (f.relative_path.as_str(), f))
            .collect();
        let map2: HashMap<&str, &VersionFileInfo> = files2
            .iter()
            .filter(|f| f.file_status != "delete")
            .map(|f| (f.relative_path.as_str(), f))
            .collect();

        let set1: HashSet<&str> = map1.keys().copied().collect();
        let set2: HashSet<&str> = map2.keys().copied().collect();

        let mut only_in_first: Vec<DiffFileEntry> = set1
            .difference(&set2)
            .map(|p| {
                let f = map1[p];
                DiffFileEntry {
                    relative_path: f.relative_path.clone(),
                    file_hash: f.file_hash.clone(),
                    file_status: f.file_status.clone(),
                }
            })
            .collect();

        let mut only_in_second: Vec<DiffFileEntry> = set2
            .difference(&set1)
            .map(|p| {
                let f = map2[p];
                DiffFileEntry {
                    relative_path: f.relative_path.clone(),
                    file_hash: f.file_hash.clone(),
                    file_status: f.file_status.clone(),
                }
            })
            .collect();

        let mut different: Vec<DiffEntry> = Vec::new();
        for path in set1.intersection(&set2) {
            let f1 = map1[path];
            let f2 = map2[path];
            if f1.file_hash != f2.file_hash || f1.file_status != f2.file_status {
                different.push(DiffEntry {
                    relative_path: f1.relative_path.clone(),
                    file_in_v1: DiffSide {
                        file_hash: f1.file_hash.clone(),
                        file_status: f1.file_status.clone(),
                    },
                    file_in_v2: DiffSide {
                        file_hash: f2.file_hash.clone(),
                        file_status: f2.file_status.clone(),
                    },
                });
            }
        }

        only_in_first.sort_by(|a, b| a.relative_path.cmp(&b.relative_path));
        only_in_second.sort_by(|a, b| a.relative_path.cmp(&b.relative_path));

        VersionDiff {
            only_in_first: if only_in_first.is_empty() { None } else { Some(only_in_first) },
            only_in_second: if only_in_second.is_empty() { None } else { Some(only_in_second) },
            different: if different.is_empty() { None } else { Some(different) },
        }
    }
}

// Set operation helpers
fn difference<'a>(a: &HashSet<&'a str>, b: &HashSet<&'a str>) -> Vec<&'a str> {
    let mut result: Vec<&str> = a.difference(b).copied().collect();
    result.sort();
    result
}

fn intersection<'a>(a: &HashSet<&'a str>, b: &HashSet<&'a str>) -> Vec<&'a str> {
    let mut result: Vec<&str> = a.intersection(b).copied().collect();
    result.sort();
    result
}